ipfix/node.c

```diff
@@ -327,6 +502,130 @@ ipfix_node_fn (vlib_main_t * vm,
+    u64 data_size = 0;
+    vec_foreach(field, set->fields) {
+      data_size = data_size + field->size;
+    }
+
+    netflow_v10_data_set_t active_set;
+    active_set.data = malloc(data_size);
+    void *ptr = (size_t)active_set.data;
+
+    vec_foreach(field, set->fields) {
+      switch (field->identifier) {
+      case sourceIPv4Address:
+        ASSERT(field->size == sizeof(ip4_address_t));
+        memcpy(ptr, &record->flow_key.src, field->size);
+        break;
+      case destinationIPv4Address:
+        ASSERT(field->size == sizeof(ip4_address_t));
+        memcpy(ptr, &record->flow_key.dst, field->size);
+        break;
+      case protocolIdentifier:
+        ASSERT(field->size == sizeof(u8));
+        memcpy(ptr, &record->flow_key.protocol, field->size);
+        break;
+      case sourceTransportPort:
+        ASSERT(field->size == sizeof(u16));
+        memcpy(ptr, &record->flow_key.src_port, field->size);
+        break;
+      case destinationTransportPort:
+        ASSERT(field->size == sizeof(u16));
+        memcpy(ptr, &record->flow_key.dst_port, field->size);
+        break;
+      case flowStartMilliseconds:
+        ASSERT(field->size == sizeof(u64));
+        memcpy(ptr, &record->flow_start, field->size);
+        break;
+      case flowEndMilliseconds:
+        ASSERT(field->size == sizeof(u64));
+        memcpy(ptr, &record->flow_end, field->size);
+        break;
+      case octetDeltaCount:
+        ASSERT(field->size == sizeof(u64));
+        memcpy(ptr, &record->octet_delta_count, field->size);
+        break;
+      case packetDeltaCount:
+        ASSERT(field->size == sizeof(u64));
+        memcpy(ptr, &record->packet_delta_count, field->size);
+        break;
+      default:
+        ASSERT(0); // Error. We don't know what this type is!
+      };
+
+      // Advance the pointer to the next field.
+      ptr = (void *)((size_t)ptr + field->size);
+    };
+
+    vec_add1(packet->sets, active_set);
+  };
+}
+
+/* Writes `packet` to `buffer`. The buffer MUST have enough space allocated to fit the entire
+ * packet.
+ *
+ * Returns number of bytes written to buffer.
+ */
+static u64 ipfix_write_v10_data_packet(vlib_buffer_t *buffer, netflow_v10_data_packet_t *packet)
+{
+  netflow_v10_data_set_t *data_set;
+  netflow_v10_template_set_t *template_set;
+  netflow_v10_field_specifier_t *field_spec;
+  netflow_v10_template_t template;
+  ipfix_make_v10_template(&template);
+
+  u64 written = 0;
+  u64 set_idx;
+  void *ptr = buffer;
+  vec_foreach_index(set_idx, template.sets) {
+    template_set = vec_elt_at_index(template.sets, set_idx);
+    data_set = vec_elt_at_index(packet->sets, set_idx);
+
+    // Calculate the length of the set.
+    size_t set_length = sizeof(netflow_v10_set_header_t);
+    vec_foreach(field_spec, template_set->fields) {
+      set_length = set_length + field_spec->size;
+    };
+
+    // Should be able to just memcopy the entire set, data 'n all.
+    memcpy(ptr, data_set, set_length);
```

Comment:
After working with this code a bit, I think this code should explicitly `memcpy` the `data` field of the `data_set`. Right now I think it is just copying the header. The header also seems to be zeroes. Here is a patch with some hacky code that fixes it partly, but I don't think it's quite right either:

```
From 68dd8365c85e405d0ffd2e611cf82efaeb56a1a6 Mon Sep 17 00:00:00 2001
From: Asumu Takikawa <asumu@igalia.com>
Date: Mon, 27 Nov 2017 17:25:07 -0800
Subject: [PATCH] Some attempts to fix flow set writing

---
 ipfix/node.c | 15 ++++++++++-----
 1 file changed, 10 insertions(+), 5 deletions(-)

diff --git a/ipfix/node.c b/ipfix/node.c
index 541dbd5..daeb52e 100644
--- a/ipfix/node.c
+++ b/ipfix/node.c
@@ -619,17 +619,22 @@ static u64 ipfix_write_v10_data_packet(void *buffer, netflow_v10_data_packet_t *
     data_set = vec_elt_at_index(packet->sets, set_idx);
 
     // Calculate the length of the set.
-    size_t set_length = sizeof(netflow_v10_set_header_t);
+    size_t header_length = sizeof(netflow_v10_set_header_t);
+    size_t data_length = 0;
     vec_foreach(field_spec, template_set->fields) {
-      set_length = set_length + field_spec->size;
+      data_length = data_length + field_spec->size;
     };
 
     // Should be able to just memcopy the entire set, data 'n all.
-    memcpy(ptr, data_set, set_length);
-    written = written + (u64)set_length;
+    data_set->header.id = htons(1);
+    data_set->header.length = htons(data_length);
+    memcpy(ptr, &data_set->header, header_length);
+    ptr = (void*)((size_t)ptr + header_length);
+    memcpy(ptr, data_set->data, data_length);
+    written = written + (u64)header_length + (u64)data_length;
 
     // Advence the pointer past the set.
-    ptr = (void *)((size_t)ptr + set_length);
+    ptr = (void *)((size_t)ptr + header_length + data_length);
   };
 
   return written;
-- 
2.11.0

```